src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineControl.xaml.cs

```diff
@@ -0,0 +1,229 @@
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.TextManager.Interop;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.DocumentOutline
+{
+    /// <summary>
+    /// Interaction logic for DocumentOutlineControl.xaml
+    /// </summary>
+    internal partial class DocumentOutlineControl : UserControl, IVsCodeWindowEvents
+    {
+        private ILanguageServiceBrokerShim LanguageServiceBroker { get; }
+
+        private IThreadingContext ThreadingContext { get; }
+
+        private IVsEditorAdaptersFactoryService EditorAdaptersFactoryService { get; }
+
+        private IVsCodeWindow CodeWindow { get; }
+
+        /// <summary>
+        /// The type of sorting to be applied to the UI model in <see cref="UpdateUIAsync"/>.
+        /// </summary>
+        private SortOption SortOption { get; set; }
+
+        /// <summary>
+        /// Queue to batch up work to do to compute the UI model. Used so we can batch up a lot of events 
+        /// and only fetch the model once for every batch.
+        /// </summary>
+        private readonly AsyncBatchingWorkQueue<bool, DocumentSymbolModel?> _computeUIModelQueue;
+
+        /// <summary>
+        /// Queue to batch up work to do to update the UI model.
+        /// </summary>
+        private readonly AsyncBatchingWorkQueue<bool, DocumentSymbolModel?> _updateUIModelQueue;
+
+        /// <summary>
+        /// Queue to batch up work to do to highlight the currently selected symbol node, expand/collapse nodes,
+        /// then update the UI.
+        /// </summary>
+        private readonly AsyncBatchingWorkQueue<ExpansionOption> _highlightAndExpandNodesQueue;
+
+        /// <summary>
+        /// Queue to batch up work to do to select code in the editor based on the current caret position.
+        /// </summary>
+        private readonly AsyncBatchingWorkQueue<DocumentSymbolItem> _jumpToContentQueue;
+
+        /// <summary>
+        /// Keeps track of the current primary and secondary text views. Should only be accessed by the UI thread.
+        /// </summary>
+        private readonly Dictionary<IVsTextView, ITextView> _trackedTextViews = new();
+
+        public DocumentOutlineControl(
+            ILanguageServiceBrokerShim languageServiceBroker,
+            IThreadingContext threadingContext,
+            IAsynchronousOperationListener asyncListener,
+            IVsEditorAdaptersFactoryService editorAdaptersFactoryService,
+            IVsCodeWindow codeWindow,
+            CancellationToken cancellationToken)
+        {
+            InitializeComponent();
+
+            LanguageServiceBroker = languageServiceBroker;
+            ThreadingContext = threadingContext;
+            EditorAdaptersFactoryService = editorAdaptersFactoryService;
+            CodeWindow = codeWindow;
+            ComEventSink.Advise<IVsCodeWindowEvents>(codeWindow, this);
+            SortOption = SortOption.Order;
```

Comment:
aside `SortOption.Order` doesn't really seem clear to me.  it's like saying `i want order things by their order` (e.g. it sounds circular).  Would `SortOption.Location` make more sense.  e.g. `i want to order these things by their location in the file`.